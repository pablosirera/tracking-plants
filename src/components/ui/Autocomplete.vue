<template>
  <div class="relative">
    <input
      v-model="itemSelected"
      v-click-outside="onClickOutsideInput"
      type="text"
      class="border py-1 px-1 rounded w-full capitalize focus:outline-none"
      @input="changeValueAutocomplete"
    />
    <div
      class="border max-h-32 overflow-y-scroll absolute w-full bg-white"
      v-if="items.length"
    >
      <span
        v-for="item in items"
        :key="item.id"
        class="block cursor-pointer capitalize py-1 px-1 hover:font-bold hover:bg-gray-200"
        @click="selectItem(item)"
      >
        {{ item.name }}
      </span>
    </div>
  </div>
</template>

<script>
export default {
  name: 'Autocomplete',
  props: {
    items: {
      type: Array,
      default: () => []
    }
  },
  data: () => ({
    itemSelected: ''
  }),
  methods: {
    changeValueAutocomplete(event) {
      const value = event.target.value
      this.$emit('input', value)
    },
    selectItem(item) {
      this.itemSelected = item.name
      this.$emit('select-item', item)
    },
    onClickOutsideInput() {
      this.$emit('clear-item')
    }
  }
}
</script>
