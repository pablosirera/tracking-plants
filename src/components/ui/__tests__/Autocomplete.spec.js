import { shallowMount } from '@vue/test-utils'
import Autocomplete from '../Autocomplete'

describe('Global Components -> Ui -> Autocomplete', () => {
  const build = propsData => {
    const options = propsData ? { propsData } : null
    const wrapper = shallowMount(Autocomplete, options)

    return {
      wrapper
    }
  }

  it('should match snapshot', () => {
    const { wrapper } = build()
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should emit input event when autocomplete input has value', () => {
    const { wrapper } = build()
    const input = wrapper.find('[data-test="autocomplete-input"]')

    input.setValue('hello')

    expect(wrapper.emitted().input).toBeTruthy()
    expect(wrapper.emitted().input[0][0]).toEqual('hello')
  })

  it('should not render items when prop items is empty', () => {
    const { wrapper } = build()
    const autocompleteWrapper = wrapper.find(
      '[data-test="wrapper-items-autocomplete"]'
    )

    expect(autocompleteWrapper.element).not.toBeDefined()
  })

  it('should render two items', () => {
    const items = [
      { id: 1, name: 'foo' },
      { id: 2, name: 'bar' }
    ]
    const { wrapper } = build({ items })
    const autocompleteItems = wrapper.findAll('[data-test="autocomplete-item"]')

    expect(autocompleteItems.length).toBe(2)
  })

  it('should emit select-item event with clicked value', () => {
    const items = [
      { id: 1, name: 'foo' },
      { id: 2, name: 'bar' }
    ]
    const { wrapper } = build({ items })
    const autocompleteItems = wrapper.findAll('[data-test="autocomplete-item"]')

    autocompleteItems.at(0).trigger('click')

    expect(wrapper.emitted()['select-item']).toBeTruthy()
    expect(wrapper.emitted()['select-item'][0][0]).toEqual({
      id: 1,
      name: 'foo'
    })
  })
})
