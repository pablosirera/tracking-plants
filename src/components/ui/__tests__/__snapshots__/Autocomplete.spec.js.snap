// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Global Components -> Ui -> Autocomplete should match snapshot 1`] = `
<div
  class="relative"
>
  <input
    class="border py-1 px-1 rounded w-full capitalize focus:outline-none"
    data-test="autocomplete-input"
    type="text"
  />
   
  <!---->
</div>
`;
